src/Microsoft.TestPlatform.Utilities/CodeCoverageRunSettingsProcessor.cs

```diff
@@ -0,0 +1,484 @@
+
+        /// <summary>
+        /// Checks if we should process the current exclusion node.
+        /// </summary>
+        /// 
+        /// <param name="node">The current exclusion node.</param>
+        /// 
+        /// <returns>
+        /// <see cref="true"/> if the node should be processed, <see cref="false"/> otherwise.
+        /// </returns>
+        private bool ShouldProcessCurrentExclusion(XmlNode node)
+        {
+            const string attributeName = "mergeDefaults";
+            const string attributeValue = "false";
+
+            foreach (XmlAttribute attribute in node.Attributes)
+            {
+                if (attribute.Name == attributeName && attribute.Value == attributeValue)
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Extracts the node specified by the current path using the provided node as root.
+        /// </summary>
+        /// 
+        /// <param name="node">The root to be used for extraction.</param>
+        /// <param name="path">The path used to specify the requested node.</param>
+        /// 
+        /// <returns>The extracted node if successful, <see cref="null"/> otherwise.</returns>
+        private XmlNode ExtractNode(XmlNode node, string path)
+        {
+            try
+            {
+                return node.SelectSingleNode(path);
+            }
+            catch (XPathException ex)
+            {
+                EqtTrace.Error(
+                    "CodeCoverageRunSettingsProcessor.ExtractNode: Cannot select single node \"{0}\".",
+                    ex.Message);
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Adds all children nodes of the current root to the current exclusion type's exclusion
+        /// rules cache.
+        /// </summary>
+        /// 
+        /// <param name="node">The root node.</param>
+        /// <param name="exclusion">The exclusion rule.</param>
+        private void AddNodes(XmlNode node, Exclusion exclusion)
+        {
+            foreach (XmlNode child in node.ChildNodes)
+            {
+                var key = child.OuterXml;
+
+                // Ignore keys that are already present in the current exclusion type's exclusion
+                // rules cache.
+                if (exclusion.ExclusionRules.ContainsKey(key))
+                {
+                    continue;
+                }
+
+                // Add the current exclusion rule to the exclusion type's cache.
+                exclusion.ExclusionRules.Add(key, child);
+            }
+        }
+
+        /// <summary>
+        /// Merges the current settings rules with the default settings rules.
+        /// </summary>
+        /// 
+        /// <param name="node">The root node.</param>
+        /// <param name="exclusion">The exclusion rule.</param>
+        private void MergeNodes(XmlNode node, Exclusion exclusion)
```

Comment:
I think whole merging can be done with something like this:


        private void MergeNodes(XmlNode currentNode, XmlNode defaultNode, Exclusion exclusion)
        {
            HashSet<string> existingExclusions = new HashSet(node.ChildNodes.Select(c => c.OuterXml));
       
            foreach (var child in defaultNode.ChildNodes)
            {
                if(!existingExclusions.Contains(child.OuterXml))
                {
                  var importedChild = currentNode.OwnerDocument.ImportNode(child, true);
                  currentNode.AppendChild(importedChild);
                }
            }
        }
        
        
By this you don't need at all method AddNodes and ExclusionRules map.

Reply:
Awesome simplification. While we're at it I completely got rid of the `Exclusion` class as it was rendered useless by this comment.